File: JavaScript/337.打家劫舍-iii.js
/*
 * @lc app=leetcode.cn id=337 lang=javascript
 *
 * [337] 打家劫舍 III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
/* 
贴个官方题解，我没解出来
https://leetcode-cn.com/problems/house-robber-iii/solution/da-jia-jie-she-iii-by-leetcode-solution/
*/
/* 
Accepted
124/124 cases passed (104 ms)
Your runtime beats 50.21 % of javascript submissions
Your memory usage beats 21.39 % of javascript submissions (43.6 MB)
*/
var rob = function (root) {
  const f = new WeakMap()
  const g = new WeakMap()

  const dfs = node => {
    if (!node) return
    dfs(node.left)
    dfs(node.right)
    // 后续遍历
    f.set(node, node.val + (g.get(node.left) || 0) + (g.get(node.right) || 0));
    g.set(node, Math.max(f.get(node.left) || 0, g.get(node.left) || 0) + Math.max(f.get(node.right) || 0, g.get(node.right) || 0));
  }
  dfs(root)
  return Math.max((f.get(root) || 0), (g.get(root) || 0))
};
// @lc code=end

File: Python3/337.打家劫舍-iii.py
#
# @lc app=leetcode.cn id=337 lang=python3
#
# [337] 打家劫舍 III
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
"""
深度优先后续遍历 官方题解
我果然还是太菜了
Accepted
124/124 cases passed (52 ms)
Your runtime beats 36.38 % of python3 submissions
Your memory usage beats 69.14 % of python3 submissions (16.9 MB)
"""

class Solution:
    def rob(self, root: TreeNode) -> int:
        def handler(node):
            if not node:
                return 0, 0
            left_select, left_no = handler(node.left)
            right_select, right_no = handler(node.right)
            return node.val + left_no + right_no, max(left_select, left_no) + max(right_select, right_no)

        return max(handler(root))
# @lc code=end
